tensorrt_llm/_torch/pyexecutor/llm_request.py

```diff
@@ -35,10 +35,199 @@
+                 prompt_len: int,
+                 max_new_tokens: int,
+                 device_memory=False,
+                 streaming=False,
+                 return_log_probs: bool = False,
+                 return_context_logits: bool = False,
+                 return_generation_logits: bool = False):
+        self._streaming = streaming
+        self._context_logits = LogitsStorage(
+            prompt_len, device_memory) if return_context_logits else None
+        self._generation_logits = LogitsStorage(
+            max_new_tokens, device_memory) if return_generation_logits else None
+        self._log_probs = LogProbStorage() if return_log_probs else None
+
+    def append_context_logits(self, context_logits: torch.Tensor):
+        if self._context_logits:
+            self._context_logits.append(context_logits)
+
+    def append_generation_logits(self, generation_logits: torch.Tensor):
+        if self._generation_logits:
+            self._generation_logits.append(generation_logits)
+
+    def append_log_probs(self, log_probs: LogProbs):
+        if self._log_probs:
+            self._log_probs.append(log_probs)
+
+    @property
+    def context_logits(self) -> torch.Tensor | None:
+        return self._context_logits and self._context_logits.get(
+            True)[:, 0]  # remove beam_width axis for context
+
+    @property
+    def generation_logits(self) -> torch.Tensor | None:
+        # Internal storage: [seq_length, beam_width, vocab_size]
+        # API expect: [beam_width, seq_length, vocab_size]
+        return self._generation_logits and self._generation_logits.get(
+            not self._streaming).transpose(0, 1)
+
+    @property
+    def log_probs(self) -> list[list[float]] | None:
+        return self._log_probs and self._log_probs.log_probs
+
+    @property
+    def cum_log_probs(self) -> list[float] | None:
+        return self._log_probs and self._log_probs.cum_log_probs
+
+
+class LlmResult:
+    """LlmResult wraps `bindings.executor.Result` but detour some features to Python implementation"""
+    py_result_properties = frozenset(
+        ('context_logits', 'generation_logits', 'log_probs', 'cum_log_probs'))
+
+    def __init__(self, result: tensorrt_llm.bindings.executor.Result,
+                 py_result: PyResult):
+        self._result = result
+        self._py_result = py_result
+
+    def __getattr__(self, item):
+        if item in self.py_result_properties:
+            return getattr(self._py_result, item)
+        return getattr(self._result, item)
+
+
+class LlmResponse:
+    """LlmResponse wraps `bindings.executor.Response` but detour some features to Python implementation"""
+
+    def __init__(self, response: tensorrt_llm.bindings.executor.Response,
+                 py_result: PyResult):
+        self._response = response
+        self._py_result = py_result
+
+    def __getstate__(self):
+        return self._response, self._py_result
+
+    def __setstate__(self, state):
```

Comment:
Please add some tests for the pickle support of such core DS., We encountered some pickle failure that breaks the e2e tests and consumes more time to investigate.

Reply:
I'm adding one in `unittest/_torch/test_return_logits.py`. Is there better location to add it?

Reply:
I noticed that the `tests/unittest/_torch/pyexecutor` directory is missing. Typically, we add tests to a module with a similar path in the tests directory. Could we create a `tests/unittest/_torch/pyexecutor/test_llm_request.py` file to include the necessary tests? @QiJune do you have any concerns about this?

Reply:
@Superjomn  I agree with you, will refine it in the next PR.